src/main/java/net/neoforged/neoforge/event/entity/EntityTeleportEvent.java

```diff
@@ -29,43 +32,86 @@
  * All children of this event are fired on the {@link NeoForge#EVENT_BUS}.<br>
  **/
 public class EntityTeleportEvent extends EntityEvent implements ICancellableEvent {
+    private GlobalVec3 target;
+    @Deprecated(since = "1.21.1", forRemoval = true)
     protected double targetX;
+    @Deprecated(since = "1.21.1", forRemoval = true)
     protected double targetY;
+    @Deprecated(since = "1.21.1", forRemoval = true)
     protected double targetZ;
 
+    @Deprecated(since = "1.21.1", forRemoval = true)
     public EntityTeleportEvent(Entity entity, double targetX, double targetY, double targetZ) {
+        this(entity, new GlobalVec3(entity.level(), targetX, targetY, targetZ));
+    }
+
+    public EntityTeleportEvent(Entity entity, GlobalVec3 target) {
         super(entity);
-        this.targetX = targetX;
-        this.targetY = targetY;
-        this.targetZ = targetZ;
+        this.target = target;
     }
 
     public double getTargetX() {
-        return targetX;
+        return target.x;
     }
 
     public void setTargetX(double targetX) {
-        this.targetX = targetX;
+        target = target.withX(targetX);
     }
 
     public double getTargetY() {
-        return targetY;
+        return target.y;
     }
 
     public void setTargetY(double targetY) {
-        this.targetY = targetY;
+        target = target.withY(targetY);
     }
 
     public double getTargetZ() {
-        return targetZ;
+        return target.z;
     }
 
     public void setTargetZ(double targetZ) {
-        this.targetZ = targetZ;
+        target = target.withZ(targetZ);
+    }
+
+    public Level getTargetLevel() {
+        return target.level;
+    }
+
+    @ApiStatus.Internal
+    // Casting helper. Backup case shouldn't ever happen, see protection in setTargetLevel().
+    public ServerLevel getTargetServerlevelOr(ServerLevel backup) {
+        return target.level instanceof ServerLevel serverlevel ? serverlevel : backup;
+    }
+
+    /**
+     * Checks if the event sub-type support changing the target level.
+     * <p>
+     * Note that a positive result is not a guarantee that the change will be honored by a modded mode of teleportation. We do not recommend changing the target level when original source and target levels are the same.
+     */
+    public boolean supportsTargetLevelChange() {
+        return true;
     }
 
+    /**
+     * Changes the target Level of the teleportation. Will throw an {@link IllegalStateException} if that is not supported or if the type of the provided level (client/server) doesn't match the original value's.
```

Comment:
Note the fact that extenders of this class must override this method and throw that exception if they do not support changing the target level (per `supportsTargetLevelChange()`) as an `@implSpec`.

Also, document the throwing of the exception as a `@throws` tag, not just in prose.